elasticdl/doc/custom_data_source_support_design.md

```diff
@@ -0,0 +1,77 @@
+## Design for Custom Data Source Support
+
+This document describes the design for supporting custom data source in ElasticDL.
+
+### Support Custom Data Source in ElasticDL
+
+There are a couple of places in the the current codebase (as of commit [77cc87a](https://github.com/sql-machine-learning/elasticdl/tree/77cc87a90eec54db565849f0ae07d271fd957190))
+that are coupled heavily with RecordIO, for example:
+
+* When we create and dispatch tasks via ``_make_task_dispatcher()``, we rely on scanning directories that contain
+RecordIO files and reading the index for each file to obtain dictionaries of `{file_path: num_records}`.
+We then use this information to create tasks.
+* ``Worker`` starts a `TaskDataService` that generates a `tf.data.Dataset` instance that contains the actual data.
+The data is fetched in `TaskDataService._gen()` by opening a RecordIO reader to read the data offsets that each task
+we previously created.
+
+In order to support custom data sources in ElasticDL, we propose to make the following changes:
+
+* Rename ``shard_file_name`` in `Task` protobuf definition to `shard_name` so a task is independent with file names and `shard_name` will
+serve as an identifier for the shard.
```

Comment:
So the new `shard` represents a file, or a table, or sth else?

Reply:
It represents a set that contains multiple records, which can be file, table, etc.